README.md
# stream_test_scheduler

Test if a stream emits elements at a certain time (window). It works similar to [TestScheduler][1] in [RxJS][2], but it uses real time (milliseconds) instead of virtual time, since you can't fake time in Dart (see [Irn's comment][3] on stackoverflow). You can pass a maximum deviation to the matcher to specify a time window.

  [1]: https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/testing/testscheduler.md
  [2]: https://github.com/Reactive-Extensions/RxJS
  [3]: http://stackoverflow.com/questions/32345853/dart-how-to-test-if-stream-emits-elements-at-a-certain-time#comment52575633_32345853

## Usage

This example tests the function `Stream delay(Stream input, Duration duration)` (from [stream_ext][stream_ext] package) that time shifts elements on a stream `input` by `duration`. The relative time intervals between the values are preserved.

    import 'dart:async';
    
    import 'package:stream_ext/stream_ext.dart';
    import 'package:stream_test_scheduler/stream_test_scheduler.dart';
    import 'package:test/test.dart';
    
    main() async {
      test('delay preserves relative time intervals between the values', () async {
        var scheduler = new TestScheduler();
    
        var source = scheduler.createStream(
            [onNext(150, 1), onNext(200, 2), onNext(230, 3), onCompleted(260)]);
    
        var result = await scheduler
            .startWithCreate(() => StreamExt.delay(source, ms(1000)));
    
        expect(
            result,
            equalsRecords([
              onNext(1150, 1),
              onNext(1200, 2),
              onNext(1230, 3),
              onCompleted(1260)
            ], maxDeviation: 20));
      });
    }

I use a maximum deviation of 20 milliseconds in this example. But deviation varies. You might have to use a different maximum deviation value for another test or on another (faster/slower) system.

  [stream_ext]: https://pub.dartlang.org/packages/stream_ext 

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

  [tracker]: https://github.com/maiermic/stream_test_scheduler/issues


CHANGELOG.md
# Changelog

## 0.3.0

- added equality check for iterable, list, set and map

## 0.2.0

- onNext takes a dynamic value instead of an int


lib/src/record.dart
// Copyright (c) 2015, Michael Maier. All rights reserved. Use of this source code
// is governed by a MIT-style license that can be found in the LICENSE file.

library stream_test_scheduler.record;

abstract class Record implements Comparable {
  final Duration ticks;
  Record(this.ticks);

  @override
  int compareTo(other) => ticks.compareTo(other.ticks);
}

class OnNextRecord extends Record {
  final value;
  OnNextRecord(Duration ticks, this.value) : super(ticks);

  @override
  String toString() => 'onNext($value)@${ticks.inMilliseconds}';
}

class OnErrorRecord extends Record {
  final exception;
  OnErrorRecord(Duration ticks, this.exception) : super(ticks);

  @override
  String toString() => 'onError($exception)@${ticks.inMilliseconds}';
}

class OnCompletedRecord extends Record {
  OnCompletedRecord(Duration ticks) : super(ticks);

  @override
  String toString() => 'onCompleted()@${ticks.inMilliseconds}';
}


lib/src/matcher.dart
// Copyright (c) 2015, Michael Maier. All rights reserved. Use of this source code
// is governed by a MIT-style license that can be found in the LICENSE file.

library stream_test_scheduler.matcher;

import 'package:matcher/matcher.dart';
import 'package:stream_test_scheduler/src/record.dart';
import 'package:collection/equality.dart';

equalsRecords(List<Record> records, {int maxDeviation: 0}) {
  return pairwiseCompare(records, (Record r1, Record r2) {
    var deviation = (r1.ticks.inMilliseconds - r2.ticks.inMilliseconds).abs();
    if (deviation > maxDeviation) {
      return false;
    }
    if (r1 is OnNextRecord && r2 is OnNextRecord) {

      if(r1.value is Iterable && r2.value is Iterable) {
        return new IterableEquality().equals(r1.value, r2.value);
      }
      if(r1.value is List && r2.value is List) {
        return new ListEquality().equals(r1.value, r2.value);
      }
      if(r1.value is Set && r2.value is Set) {
        return new SetEquality().equals(r1.value, r2.value);
      }
      if(r1.value is Map && r2.value is Map) {
        return new MapEquality().equals(r1.value, r2.value);
      }

      return r1.value == r2.value;
    }
    if (r1 is OnErrorRecord && r2 is OnErrorRecord) {
      return r1.exception == r2.exception;
    }
    return (r1 is OnCompletedRecord && r2 is OnCompletedRecord);
  }, 'equal with deviation of ${maxDeviation}ms to');
}


lib/src/scheduler.dart
// Copyright (c) 2015, Michael Maier. All rights reserved. Use of this source code
// is governed by a MIT-style license that can be found in the LICENSE file.

library stream_test_scheduler.scheduler;

import 'dart:async';

import 'package:stream_test_scheduler/src/record.dart';
import 'package:stream_test_scheduler/src/record_aux.dart';

class TestScheduler {
  final _SchedulerTasks _tasks;

  TestScheduler() : _tasks = new _SchedulerTasks();

  Stream createStream(List<Record> records) {
    final controller = new StreamController(sync: true);
    _tasks.add(controller, records);
    return controller.stream;
  }

  Future<List<Record>> startWithCreate(Stream createStream()) {
    final completer = new Completer<List<Record>>();
    final records = <Record>[];
    final start = new DateTime.now();
    int timeStamp() {
      final current = new DateTime.now();
      return current.difference(start).inMilliseconds;
    }
    createStream().listen((event) => records.add(onNext(timeStamp(), event)),
        onError: (exception) => records.add(onError(timeStamp(), exception)),
        onDone: () {
      records.add(onCompleted(timeStamp()));
      completer.complete(records);
    });
    _tasks.run();
    return completer.future;
  }
}

class _SchedulerTasks {
  Map<Record, StreamController> _controllers = {};
  List<Record> _records = [];

  void add(StreamController controller, List<Record> records) {
    for (var record in records) {
      _controllers[record] = controller;
    }
    _records.addAll(records);
  }

  void run() {
    _records.sort();
    for (var record in _records) {
      final controller = _controllers[record];
      new Future.delayed(record.ticks, () {
        if (record is OnNextRecord) {
          controller.add(record.value);
        } else if (record is OnErrorRecord) {
          controller.addError(record.exception);
        } else if (record is OnCompletedRecord) {
          controller.close();
        }
      });
    }
  }
}


example/stream_test_scheduler.dart
// Copyright (c) 2015, Michael Maier. All rights reserved. Use of this source code
// is governed by a MIT-style license that can be found in the LICENSE file.

library stream_test_scheduler.example;

import 'package:stream_ext/stream_ext.dart';
import 'package:stream_test_scheduler/stream_test_scheduler.dart';
import 'package:test/test.dart';

main() async {
  test('delay preserves relative time intervals between the values', () async {
    var scheduler = new TestScheduler();

    var source = scheduler.createStream(
        [onNext(150, 1), onNext(200, 2), onNext(230, 3), onCompleted(260)]);

    var result = await scheduler
        .startWithCreate(() => StreamExt.delay(source, ms(1000)));

    expect(
        result,
        equalsRecords([
          onNext(1150, 1),
          onNext(1200, 2),
          onNext(1230, 3),
          onCompleted(1260)
        ], maxDeviation: 20));
  });
}
